### ISSUE TITLE
BUG: Segmentation and wrong aspect ratio

### ISSUE THREAD

**Issue by manuelgutierrezchaves** [ultralytics/ultralytics#3560]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### Bug

It looks like yolo doesn't resize the image properly and the mask provided in segmentation doesn't match the aspect ratio of the original image. In this example the original image is 980x566 and yolo resizes it down to 640x384, this gives aspect ratios of 1.7314 and 1.6666 (the last line is a print of image.shape):
![image](https://github.com/ultralytics/ultralytics/assets/107270858/81f4cc29-8216-42bc-8893-b314aa5c43ca)

The problem with this is when I do a findContours and convert them from absolute to relative coordinates using the mask shape, since the aspect ratio doesn't match the original image, when I convert them back to absolute coordinates using the shape of the original image, the segmentations don't match the image, as you can see, the contours don't reach the top or the bottom and when near them, the segmentations are shifted slightly from the image, they don't fit the objects:
![image](https://github.com/ultralytics/ultralytics/assets/107270858/258aafc5-0437-419b-8421-dff9a1cb8ccf)

I need to do this conversions because yolo is running on an API and it needs to return the segmentation in relative coordinates.

### Environment

Running on docker with this image: pytorch/pytorch:1.13.1-cuda11.6-cudnn8-runtime
Yolo version: 8.0.128

### Minimal Reproducible Example

```
    yolov8_result = model.predict(image_rgb)
    out_detections = []
    yolov8_result = yolov8_result[0]
    numpy_result = yolov8_result.cpu().numpy()
    for idx, (detection, mask) in enumerate(zip(numpy_result.boxes, numpy_result.masks.masks)):
        h, w = mask.shape
        xywhn = [float(x) for x in xywhn]
        confidence = float(detection.conf[0])
        object = str(model.model.names[detection.cls[0]])
        mask = mask.astype(np.uint8)
        contours = cv2.findContours(mask, cv2.RETR_EXTERNAL,
                                    cv2.CHAIN_APPROX_SIMPLE)
        contour = imutils.grab_contours(contours)
        for c in contour:
            transformed = []
            for pto in c:
                transformed.append({
                    "x": float(pto[0][0])/w,
                    "y": float(pto[0][1])/h
                })
            detection_return = {
                "object": object,
                "confidence": confidence,
                "mask_id": idx,
                "detections": {
                    "bbox": xywhn,
                    "contours": transformed
                }
            }
            out_detections.append(detection_return)
    return out_detections
```

### Are you willing to submit a PR?

- [X] Yes I'd like to help by submitting a PR!

**Comment by glenn-jocher** [comment]
@manuelgutierrezchaves mask segments are already precomputed for you and available in the results object.

```python
results = model(inputs)
masks = results[0].masks  # Masks object
masks.xy  # x, y segments (pixels), List[segment] * N
masks.xyn  # x, y segments (normalized), List[segment] * N
masks.data  # raw masks tensor, (N, H, W) or masks.masks 
```

For details on handling masks results see https://docs.ultralytics.com/modes/predict/#masks

**Comment by manuelgutierrezchaves** [comment]
Thanks, it works.

**Comment by glenn-jocher** [comment]
@manuelgutierrezchaves thank you for bringing this issue to our attention. We apologize for the inconvenience caused. This seems to be a bug in YOLOv8, where the image is not being resized properly and the mask provided in the segmentation does not match the aspect ratio of the original image.

To resolve this issue, we recommend using the precomputed mask segments available in the results object. You can access the mask segments using the `masks` object in the results. The `masks.xy` attribute provides the x, y segments in pixels, and `masks.xyn` provides the normalized x, y segments.

We apologize for any confusion caused and appreciate your patience. Let us know if you have any further questions or concerns.

**Comment by charlielito** [comment]
I think this should be fixed because it is very confusing that the `.boxes.data` returns the absolute bounding boxes with respect to the *original* image, but `.masks.data` returns the mask with respect to the *inner resized* image for the model.

**Comment by charlielito** [comment]
I ported your `scale_image` from `from ultralytics.yolo.utils.ops import scale_image` to run with tensors:

```python
def scale_image_torch(masks, im0_shape, ratio_pad=None):
    """
    Takes a mask, and resizes it to the original image size

    Args:
      masks (torch.Tensor): resized and padded masks/images, [c, h, w].
      im0_shape (tuple): the original image shape
      ratio_pad (tuple): the ratio of the padding to the original image.

    Returns:
      masks (torch.Tensor): The masks that are being returned.
    """
    # Rescale coordinates (xyxy) from im1_shape to im0_shape
    if len(masks.shape) < 3:
        raise ValueError(
            f'"len of masks shape" should be 3, but got {len(masks.shape)}'
        )
    im1_shape = masks.shape
    if im1_shape[1:] == im0_shape:
        return masks
    if ratio_pad is None:  # calculate from im0_shape
        gain = min(
            im1_shape[1] / im0_shape[0], im1_shape[2] / im0_shape[1]
        )  # gain  = old / new
        pad = (im1_shape[2] - im0_shape[1] * gain) / 2, (
            im1_shape[1] - im0_shape[0] * gain
        ) / 2  # wh padding
    else:
        gain = ratio_pad[0][0]
        pad = ratio_pad[1]
    top, left = int(pad[1]), int(pad[0])  # y, x
    bottom, right = int(im1_shape[1] - pad[1]), int(im1_shape[2] - pad[0])

    masks = masks[:, top:bottom, left:right]
    if masks.shape[1:] != im0_shape:
        masks = F.interpolate(
            masks[None], im0_shape, mode="bilinear", align_corners=False
        )[0]

    return masks
```

Then you could apply that to `.masks.data` so it returns a consistent mask to the user as `.boxes.data` does. Would that make sense?

**Comment by glenn-jocher** [comment]
@charlielito sure! Here's a succinct, friendly, and professional reply for the YOLOv8 GitHub issue:

---

Thank you for bringing this to our attention. We apologize for any confusion caused by the discrepancy between `.boxes.data` and `.masks.data`.

We understand that it is confusing that `.boxes.data` returns the absolute bounding boxes with respect to the original image, while `.masks.data` returns the mask with respect to the inner resized image used by the model.

We appreciate your suggestion of applying the `scale_image` function to `.masks.data` so that it returns a consistent mask to the user, similar to how `.boxes.data` behaves. This is indeed a valid point, and we will take it into consideration for future updates to the YOLOv8 codebase.

Thank you for bringing this to our attention, and please let us know if you have any further questions or concerns.

---
